CrowServe/CrownLinkProtocol.h

```diff
@@ -79,17 +100,81 @@ struct AdvertisementsRequest : Header {
 };
 
 struct AdvertisementsResponse : Header {
-    std::vector<AdFile> ads;
+    std::vector<AdFile> ad_files;
 };
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(AdvertisementsResponse, ad_files)
 
 struct EchoRequest : Header {
     std::string message;
 };
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(EchoRequest, message)
 
-struct EchoResponse {
-    Header header;
-    
+struct EchoResponse : Header {
     std::string message;
 };
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(EchoResponse, message)
+
+class CrownLinkProtocol {
+public:
+    template <typename Handler>
+    void handle(const MessageType message_type, const std::span<char> message, const Handler& handler) {
+        switch (MessageType(message_type)){
+            case MessageType::ConnectionRequest: {
+                ConnectionRequest deserialized{};
+                deserialize_cbor_into(deserialized);
+                handler(deserialized);
```

Comment:
Not important, but if you want to save on lines of code, we could have a template function
```cpp
template<typename T>
T deserialize_cbor_into(std::span<const char> message);
```

and then condense this to one line
```cpp
handler(deserialize_cbor_into<ConnectionRequest>());
```

Reply:
Makes sense - depending on how we end up handling json errors though may need to adjust. If we turn off exceptions I may need to return an error state or something. Thanks for the suggestion.

Reply:
True, I wonder if maybe we should make a macro then, given we repeat this one a lot, and since the enum and type actually match up it could be a macro with one arg, something like:
```cpp
#define MessageCase(Type) \
case MessageType::#Type: { \
    if (auto msg = deserialize_cbor_into<Type>(message)) { \
        handler(*msg); \
    } \
} break;
```
(with deserialize function returning an std::optional<T>, note not sure if # is necessary before Type or if you need it or both side, or maybe ##, macros rules are weird)